### PR TITLE
Refactor analysis output metrics: part 4 Dataclass helpers

### PR DESCRIPTION
**Pull request by dvadym**
Those helpers will be used in the following PRs

### CODE REVIEW

File: analysis/cross_partition_combiners.py [hunk 1]
```diff
@@ -94,3 +93,56 @@ def _sum_metrics_to_metric_utility(
...
+def _add_dataclasses_by_fields(dataclass1, dataclass2,
+                               fields_to_ignore: List[str]) -> None:
+    """Recursively adds all numerical fields of one dataclass to another.
+
+    The result is in dataclass1.
+
+    Assumptions:
+      1. dataclass1 and dataclass2 are instances of the same dataclass type.
+      2. all fields which should be processed (i.e. not in fields_to_ignore)
+       are dataclasses or '+' operator can be applied.
```
Comment (miracvbasaran) [inline analysis/cross_partition_combiners.py:117]: nit: `are either dataclasses or support `+` operator.`

File: analysis/cross_partition_combiners.py [hunk 2]
```diff
@@ -94,3 +93,56 @@ def _sum_metrics_to_metric_utility(
...
+        dropped_partitions=metrics(mean=prob_keep,
+                                   var=prob_keep * (1 - prob_keep)),
+        ratio_dropped_data=0)  # todo: implement
+
+
+def _add_dataclasses_by_fields(dataclass1, dataclass2,
+                               fields_to_ignore: List[str]) -> None:
+    """Recursively adds all numerical fields of one dataclass to another.
+
+    The result is in dataclass1.
```
Comment (miracvbasaran) [inline analysis/cross_partition_combiners.py:112]: nit: `The result is stored in dataclass1. dataclass2 is unmodified.

File: analysis/cross_partition_combiners.py [hunk 3]
```diff
@@ -94,3 +93,56 @@ def _sum_metrics_to_metric_utility(
...
+      1. dataclass1 and dataclass2 are instances of the same dataclass type.
+      2. all fields which should be processed (i.e. not in fields_to_ignore)
+       are dataclasses or '+' operator can be applied.
+      3. For all dataclasses fields assumptions 1,2 apply.
+
+    Attributes:
+        dataclass1, dataclass2: instances of the same dataclass type.
+        fields_to_ignore: field names which should be ignored.
+    """
+    assert type(dataclass1) == type(dataclass2)
```
Comment (miracvbasaran) [inline analysis/cross_partition_combiners.py:124]: I think we might need an error message here

File: analysis/cross_partition_combiners.py [hunk 4]
```diff
@@ -94,3 +93,56 @@ def _sum_metrics_to_metric_utility(
...
+        value1 = getattr(dataclass1, field.name)
+        value2 = getattr(dataclass2, field.name)
+        if dataclasses.is_dataclass(value1):
+            _add_dataclasses_by_fields(value1, value2, fields_to_ignore)
+            continue
+        setattr(dataclass1, field.name, value1 + value2)
+
+
+def _multiply_float_dataclasses_field(dataclass, factor: float):
+    """Recursively multiply all float fields of the dataclass by given number."""
```
Comment (miracvbasaran) [inline analysis/cross_partition_combiners.py:138]: nit: Maybe mention that it modifies `dataclass`?

File: analysis/cross_partition_combiners.py [hunk 5]
```diff
@@ -94,3 +93,56 @@ def _sum_metrics_to_metric_utility(
...
+    fields = dataclasses.fields(dataclass)
+    for field in fields:
+        value = getattr(dataclass, field.name)
+        if value is None:
+            continue
+        if field.type is float:
+            setattr(dataclass, field.name, value * factor)
+        if dataclasses.is_dataclass(value):
+            _multiply_float_dataclasses_field(value, factor)
+            continue
```
Comment (miracvbasaran) [inline analysis/cross_partition_combiners.py:148]: Optional:
```python
if value is not None and field.type is float:
    setattr(dataclass, field.name, value * factor)
if dataclasses.is_dataclass(value):
    _multiply_float_dataclasses_field(value, factor)
```

I don't think you need the continue at the end.

Also, what happens if you do `None * factor`? Maybe it is handled gracefully and results in `None`?

File: analysis/tests/cross_partition_combiners_test.py [hunk 6]
```diff
@@ -71,5 +72,46 @@ def test_metric_utility_empty_partition(self):
         self.assertEqual(output.num_empty_partitions, 1)
 
 
+# Dataclasses for DataclassHelpersTests
+@dataclasses.dataclass
+class dataclass1:
```
Comment (miracvbasaran) [inline analysis/tests/cross_partition_combiners_test.py:77]: nit: I think more descriptive names would be helpful. E.g. `flatdataclass` and `nesteddataclass` perhaps

**Review by miracvbasaran — APPROVED** [review]

File: analysis/cross_partition_combiners.py (hunk 3 above)
Comment (dvadym) [reply analysis/cross_partition_combiners.py]: done

File: analysis/cross_partition_combiners.py (hunk 5 above)
Comment (dvadym) [reply analysis/cross_partition_combiners.py]: Thx, we don't need, it was left after some refactroing ))

`None * factor` throws errors

File: analysis/tests/cross_partition_combiners_test.py (hunk 6 above)
Comment (dvadym) [reply analysis/tests/cross_partition_combiners_test.py]: good point, I named `OuterClass` and `InnerClass`

**Review by dvadym — COMMENTED** [review]
Thanks for review!

**Pull request merged**